hts.1
.TH HTS 1
.SH NAME
hts \- httptunnel server
.SH SYNOPSIS
.B hts
.I "[options] [host:][port]"
.SH "DESCRIPTION"
This manual page briefly documents the
.BR hts
command.
.PP
.B hts
listens for incoming httptunnel connections at PORT (default port is
8888), and optionally binds to ip address HOST.
When a connection is made, I/O is redirected to the destination specified
by the \-\-device or \-\-forward\-port switch.
.SH OPTIONS
The program follows the usual GNU command line syntax, with long
options starting with two dashes (`\-').
A summary of options are included below.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-c, \-\-content-length BYTES
use HTTP PUT requests of BYTES size (k, M, and G postfixes recognized)
.TP
.B \-d, \-\-device DEVICE
use DEVICE for input and output
.TP
.B \-F, \-\-forward\-port HOST:PORT
connect to PORT at HOST and use it for input and output
.TP
.B \-k, \-\-keep\-alive SECONDS
send keepalive bytes every SECONDS seconds (default is 5)
.TP
.B \-M, \-\-max\-connection\-age SEC
maximum time a connection will stay open is SEC seconds (default is 300)
.TP
.B \-s, \-\-stdin\-stdout
use stdin/stdout for communication (implies \-\-no\-daemon)
.TP
.B \-S, \-\-strict\-content\-length
always write Content-Length bytes in requests
.TP
.B \-V, \-\-version
output version information and exit
.TP
.B \-w, \-\-no-daemon
don't fork into the background
.TP
.B \-p, \-\-pid\-file LOCATION
write a PID file to LOCATION
.TP
.B \-C, \-\-chroot LOCATION
chroot to LOCATION before serving clients
.TP
.B \-u, \-\-user USER
change user and group identities before serving clients
.SH AUTHOR
This manual page was contributed by Teemu Hukkanen <tjhukkan@iki.fi>,
and was originally written for the Debian GNU/Linux system.


NEWS
* Version 4.0

Summary of user-visible changes since 3.0:

** Fixed bugs.

* Version 3.0

Summary of user-visible changes since 2.0:

** Bug fixes and improved stability.

** Long options work on all platforms.

   The GNU libc getopt_long() functions were copied into the httptunnel
   distribution and is used on systems where no long options are available
   in the standard libraries.

** Logging messages get a time stamp.

** httptunnel is now Debianized.

** Replaced the MANUAL file with two man pages.

** Made compatible with the Univeral TUN device driver.

** New --strict-content-length option.

   This option makes httptunnel obey the Content-Length field in the
   HTTP header.  This improves compatibility with some proxies.  The
   defalt is to be sloppy about the Content-Length field.

** New --keep-alive option.

   With this option, httptunnel will send keep-alive bytes, which
   improves stability with some proxies.  The defailt is to send
   keep-alive bytes every five seconds.

** New --max-connection-age option.

   This option specifies the maximum age of any HTTP connection made
   by httptunnel.  The default is 300 seconds.

** New --proxy-authorization option.

   This option is used to send a authorization string to the proxy.

** New --proxy-authorization-file option.

   This option works as --proxy-authorization, only it reads the
   string from the specified file.

** New --pid-file option.

   With this option, httptunnel will write its process identification
   number to the specified file.

** New --user-agent option.

   Use this option to specify a string for the User-Agent field sent
   in HTTP headers.  This is useful for proxies that want a specific
   value there.

* Version 2.0

Summary of user-visible changes since 1.0:

** There is a FAQ.

** There is a MANUAL which isn't.

** Port numbers have default values.

** There is a debug mode.

   Using --enable-debug with 'configure' will make htc and hts recognize
   --debug and --logfile options.

** Uses syslog.

   When not in debug mode, important events and unexpected errors
   are logged using LOG_NOTICE and LOG_ERROR, respectively.

** hts takes a --content-length option.

   It works like the one for htc.

** Short option for --forward-port changed from -f to -F.

** Improved portability.

   Works on at least some versions of GNU/Linux, Solaris, AIX,
   Digital UNIX, Cygwin, FreeBSD, OpenBSD.

** Improved protocol and HTTP handling.

   Resulting in better stability.

** Bug fixes.

** Bug reports should now go to bug-httptunnel@nocrew.org.

* Version 1.103

** Bug reporting email addres changed to bug-httptunnel@gnu.org.

** Miscellaneous cosmetic stuff.

* Version 1.102

** Emulate endprotoent() if not present.

** Changed email addresses to preferred ones.

** Handle setsockopt() errors slightly more gracefully.

* Version 1.101

** Disabled compilation of tt and rw, as they are hardly needed anymore.

** Changed internal parameters once again.

   Send keep-alive bytes ever 5 seconds.

   Maximum connection time back to 5 minutes.

** Improved logging.

   Write package name, package version and debug level to log.

   Removed one superfluous level of indentation.

** Removed a problem which caused a busy loop.

   This only happened when --proxy-buffer-* was active.

* Version 1.100

** Removed an attempt to printf() NULL, which caused a crash on Solaris.

   Thanks to Tomas Berndtsson <tomas@nocrew.org> for the fix.

** Hopefully the setsockopt() errors will be gone.

** Changed some internal parameters to satisfy really picky proxies.

   Increased frequency of keep-alive bytes from 1/30s to 1/10s.

   Decreased maximum connection time from 5 minutes to 1 minute.

   TODO: user interface for these parameters.

* Version 1.99

** Added plea for post cards to README.

   The greedy author tries to turn this otherwise perfectly free
   software into a vehicle for extracting post cards from innocent
   users.

** Fixed --enable-debug without --debug bug.

** Adjusted debugging levels.

   0 now means no debugging messages, 5 is the highest level.

* Version 1.98

** Close a HTTP POST connection after 5 minutes.

   This is a kludge, but finally no data is lost with Squid 2.

** httptunnel is now GNU software.

** Changed the short option for --forward-port from -f to -F.

   Because I want to use -f for a future --file option.

** Removed some possible buffer overflows.

* Version 1.97

** Improved poll() emulation.

** Improved HTTP request and response parsing.

** hts --content-length works now.

   Thanks Philip!

** Try some setsockopt() settings to make connections more reliable.

* Version 1.96

** Changed compiler warnings when --enable-debug.

** Included the HTTP RFCs in the distribution.

** Emulate poll() using select() on systems without poll().

   Still untested, though.

** Fixed void * arithmetic and a prototype/definition mismatch.

   Thanks to Larry Gensch for this patch!

* Version 1.95

** I/O byte counters.

   --debug 3 will show byte counters for the data sent and recieived.
   --debug 4 will also show byte counters for the raw protocol data
   sent and received.

** hts takes a --content-length switch.

** Make sure HTTP GET responses are not cached.

   Added "Expires: 0" (invalid HTTP, I know) and "Cache-Control:
   no-store, must-revalidate" to HTTP GET response header.

* Version 1.94

** Avoid a few compilation warnings.

** http.h didn't make it into the distribution.

* Version 1.93

** Parse HTTP GET responses.

   If the version is unknown or the status code signals an error, fail.

* Version 1.92

** Forgot to regenerate configure, Makefile, etc.

* Version 1.91

** Forgot to update NEWS before releasing 1.90.

* Version 1.90

** More robust.

** Major protocol change.

   Now uses the protocol for HTTP GET requests too.

** Improved error logging.

** Solaris and AIX fixes.

   Thanks to Andrew Gray for the patch.


hts.c
/*
hts.c

Copyright (C) 1999, 2000 Lars Brinkhoff.  See COPYING for terms and conditions.

hts is the server half of httptunnel.  httptunnel creates a virtual
two-way data path tunneled in HTTP requests.
*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd_.h>
#include <signal.h>
#include <sys/poll_.h>
#include <sys/time.h>
#include <sys/types.h>
#include <pwd.h>
#include <grp.h>
#include <time.h>

#include "common.h"

typedef struct
{
  char *me;
  char *device;
  const char *host;
  int port;
  char *forward_host;
  int forward_port;
  size_t content_length;
  char *pid_filename;
  int use_std;
  int use_daemon;
  int strict_content_length;
  int keep_alive;
  int max_connection_age;
  char *root;
  char *user;
} Arguments;

int debug_level = 0;
FILE *debug_file = NULL;

static void
usage (FILE *f, const char *me)
{
  fprintf (f,
"Usage: %s [OPTION]... [HOST:][PORT]\n"
"Listen for incoming httptunnel connections at PORT (default port is %d).\n"
"When a connection is made, I/O is redirected to the destination specified\n"
"by the --device, --forward-port or --stdin-stdout switch.\n"
"\n"
"  -c, --content-length BYTES     use HTTP PUT requests of BYTES size\n"
"                                 (k, M, and G postfixes recognized)\n"
"  -d, --device DEVICE            use DEVICE for input and output\n"
#ifdef DEBUG_MODE
"  -D, --debug [LEVEL]            enable debug mode\n"
#endif
"  -F, --forward-port HOST:PORT   connect to PORT at HOST and use it for \n"
"                                 input and output\n"
"  -h, --help                     display this help and exit\n"
"  -k, --keep-alive SECONDS       send keepalive bytes every SECONDS seconds\n"
"                                 (default is %d)\n"
#ifdef DEBUG_MODE
"  -l, --logfile FILE             specify logfile for debug output\n"
#endif
"  -M, --max-connection-age SEC   maximum time a connection will stay\n"
"                                 open is SEC seconds (default is %d)\n"
"  -r, --chroot ROOT              change root to ROOT\n"
"  -s, --stdin-stdout             use stdin/stdout for communication\n"
"                                 (implies --no-daemon)\n"
"  -S, --strict-content-length    always write Content-Length bytes in requests\n"
"  -u, --user USER                change user to USER\n"
"  -V, --version                  output version information and exit\n"
"  -w, --no-daemon                don't fork into the background\n"
"  -p, --pid-file LOCATION        write a PID file to LOCATION\n"
"\n"
"Report bugs to %s.\n",
	   me, DEFAULT_HOST_PORT, DEFAULT_KEEP_ALIVE,
	   DEFAULT_MAX_CONNECTION_AGE, BUG_REPORT_EMAIL);
}

static void
parse_arguments (int argc, char **argv, Arguments *arg)
{
  int c;

  /* defaults */

  arg->me = argv[0];
  arg->host = NULL;
  arg->port = DEFAULT_HOST_PORT;
  arg->device = NULL;
  arg->forward_host = NULL;
  arg->forward_port = -1;
  arg->content_length = DEFAULT_CONTENT_LENGTH;
  arg->pid_filename = NULL;
  arg->use_std = FALSE;
  arg->use_daemon = TRUE;
  arg->strict_content_length = FALSE;
  arg->keep_alive = DEFAULT_KEEP_ALIVE;
  arg->max_connection_age = DEFAULT_CONNECTION_MAX_TIME;
  arg->user = NULL;
  arg->root = NULL;
  
  for (;;)
    {
      int option_index = 0;
      static struct option long_options[] =
      {
	{ "help", no_argument, 0, 'h' },
	{ "version", no_argument, 0, 'V' },
	{ "no-daemon", no_argument, 0, 'w' },
	{ "user", required_argument, 0, 'u' },
#ifdef DEBUG_MODE
	{ "debug", required_argument, 0, 'D' },
	{ "logfile", required_argument, 0, 'l' },
#endif
	{ "chroot", required_argument, 0, 'r' },
	{ "stdin-stdout", no_argument, 0, 's' },
	{ "device", required_argument, 0, 'd' },
	{ "pid-file", required_argument, 0, 'p' },
	{ "keep-alive", required_argument, 0, 'k' },
	{ "forward-port", required_argument, 0, 'F' },
	{ "content-length", required_argument, 0, 'c' },
	{ "strict-content-length", no_argument, 0, 'S' },
	{ "max-connection-age", required_argument, 0, 'M' },
	{ 0, 0, 0, 0 }
      };

      static const char *short_options = "c:d:F:hk:M:p:sSVwu:r:"
#ifdef DEBUG_MODE
	"D:l:"
#endif
	;

      c = getopt_long (argc, argv, short_options,
		       long_options, &option_index);
      if (c == -1)
	break;

      switch (c)
	{
	case 0:
	  fprintf (stderr, "option %s", long_options[option_index].name);
	  if (optarg)
	    fprintf (stderr, " with arg %s", optarg);
	  fprintf (stderr, "\n");
	  break;
	  
	case 'c':
	  arg->content_length = atoi_with_postfix (optarg);
	  break;

	case 'd':
	  arg->device = optarg;
	  break;

#ifdef DEBUG_MODE
	case 'D':
	  if (optarg)
	    debug_level = atoi (optarg);
	  else
	    debug_level = 1;
	  break;

	case 'l':
	  debug_file = fopen (optarg, "w");
	  if (debug_file == NULL)
	    {
	      fprintf (stderr, "%s: couldn't open file %s for writing\n",
		       arg->me, optarg);
	      log_exit (1);
	    }
	  break;
#endif /* DEBUG_MODE */

	case 'F':
	  name_and_port (optarg, &arg->forward_host, &arg->forward_port);
	  if (arg->forward_port == -1)
	    {
	      fprintf (stderr, "%s: you must specify a port number.\n"
		               "%s: try '%s --help' for help.\n",
		       arg->me, arg->me, arg->me);
	      exit (1);
	    }
	  break;

	case 'h':
	  usage (stdout, arg->me);
	  exit (0);

	case 'k':
	  arg->keep_alive = atoi (optarg);
	  break;

	case 'M':
	  arg->max_connection_age = atoi (optarg);
	  break;

	case 'r':
	  arg->root = optarg;
	  break;

	case 's':
	  arg->use_std=TRUE;
	  arg->use_daemon=FALSE;
	  break;

	case 'S':
	  arg->strict_content_length = TRUE;
	  break;

	case 'u':
	  arg->user = optarg;
	  break;

	case 'V':
	  printf ("hts (%s) %s\n", PACKAGE, VERSION);
	  exit (0);

	case 'p':
	  arg->pid_filename = optarg;
	  break;

	case 'w':
	  arg->use_daemon = FALSE;
	  break;

	case '?':
	  break;

	default:
	  fprintf (stderr, "?? getopt returned character code 0%o ??\n", c);
	}
    }

  if (argc - 1 == optind)
    {
      char *colon = strchr (argv[optind], ':');

      if (colon)
	{
	  *colon = '\0';
	  arg->host = argv[optind];
	  arg->port = atoi (colon + 1);
	}
      else
	arg->port = atoi (argv[optind]);
    }
  else if (argc - 1 > optind)
    {
      usage (stderr, arg->me);
      exit (1);
    }

  if (arg->device == NULL && arg->forward_port == -1 && !arg->use_std)
    {
      fprintf (stderr, "%s: one of --device, --forward-port or --stdin-stdout must be used.\n"
	               "%s: try '%s -help' for help.\n",
	       arg->me, arg->me, arg->me);
      exit (1);
    }

  if ((arg->device != NULL && arg->forward_port != -1) ||
      (arg->device != NULL && arg->use_std) ||
      (arg->forward_port != -1 && arg->use_std))
    {
      fprintf (stderr, "%s: only one of --device, --forward-port or --stdin-stdout can be used.\n"
	               "%s: try '%s --help' for help.\n",
	       arg->me, arg->me, arg->me);
      exit (1);
    }

  if (debug_level == 0 && debug_file != NULL)
    {
      fprintf (stderr, "%s: --logfile can't be used without debugging\n",
	       arg->me);
      exit (1);
    }

  /* Removed test ((arg->device == NULL) == (arg->forward_port == -1))
   * by Sampo Niskanen - those have been tested already! */
  if (arg->port == -1 ||
      ((arg->forward_host == NULL) != (arg->forward_port == -1)))
    {
      usage (stderr, arg->me);
      exit (1);
    }
}

int
main (int argc, char **argv)
{
  int closed;
  int fd = -1;
  Arguments arg;
  Tunnel *tunnel;
  FILE *pid_file;
  uid_t uid;
  gid_t gid;

  parse_arguments (argc, argv, &arg);

  if ((debug_level == 0 || debug_file != NULL) && arg.use_daemon)
    daemon (0, 1);

#ifdef DEBUG_MODE
  if (debug_level != 0 && debug_file == NULL)
    debug_file = stderr;
#else
  openlog ("hts", LOG_PID, LOG_DAEMON);
#endif

  log_notice ("hts (%s) %s started with arguments:", PACKAGE, VERSION);
  log_notice ("  me = %s", arg.me);
  log_notice ("  device = %s", arg.device ? arg.device : "(null)");
  if (arg.host)
    log_notice ("  port = %s:%d", arg.host, arg.port);
  else
    log_notice ("  port = %d", arg.port);
  log_notice ("  forward_port = %d", arg.forward_port);
  log_notice ("  forward_host = %s",
	      arg.forward_host ? arg.forward_host : "(null)");
  log_notice ("  content_length = %d", arg.content_length);
  log_notice ("  strict_content_length = %d", arg.strict_content_length);
  log_notice ("  use_std = %d", arg.use_std);
  log_notice ("  debug_level = %d", debug_level);
  log_notice ("  pid_filename = %s",
	      arg.pid_filename ? arg.pid_filename : "(null)");
  log_notice ("  chroot = %s", arg.root ? arg.root : "(null)");
  log_notice ("  user = %s", arg.user ? arg.user : "(null)");

  tunnel = tunnel_new_server (arg.host, arg.port, arg.content_length);
  if (tunnel == NULL)
    {
      log_error ("couldn't create tunnel");
      log_exit (1);
    }

  if (tunnel_setopt (tunnel, "strict_content_length",
		     &arg.strict_content_length) == -1)
    log_debug ("tunnel_setopt strict_content_length error: %s",
	       strerror (errno));

  if (tunnel_setopt (tunnel, "keep_alive",
		     &arg.keep_alive) == -1)
    log_debug ("tunnel_setopt keep_alive error: %s", strerror (errno));

  if (tunnel_setopt (tunnel, "max_connection_age",
		     &arg.max_connection_age) == -1)
    log_debug ("tunnel_setopt max_connection_age error: %s", strerror (errno));

#ifdef DEBUG_MODE
  signal (SIGPIPE, log_sigpipe);
#else
  signal (SIGPIPE, SIG_IGN);
#endif

  if (arg.pid_filename != NULL)
    {
      pid_file = fopen (arg.pid_filename, "w+");
      if (pid_file == NULL)
        {
          fprintf (stderr, "Couldn't open pid file %s: %s\n",
		   arg.pid_filename, strerror (errno));
        }
      else
	{
          fprintf (pid_file, "%d\n", (int)getpid ());
	  if (fclose (pid_file))
            {
              fprintf (stderr, "Error closing pid file: %s\n", 
		       strerror (errno));
            }
         }
     }

  /* If requested to change user, get new uid and gid before chroot so we */
  /* don't need /etc/passwd & company in the chroot jail */
  if (arg.user)
    {
      struct passwd *pwd = getpwnam (arg.user);
      if (pwd == NULL)
        {
          log_error ("couldn't find user");
          log_exit (1);
	}
      uid = pwd->pw_uid;
      gid = pwd->pw_gid;
      if (setgroups (0, (const gid_t *)0) < 0)
        {
          log_error ("couldn't drop supplementary groups privileges");
          log_exit (1);
	}
      if (setgid (gid) < 0)
        {
          log_error ("couldn't change primary group");
          log_exit (1);
	}
      if (initgroups (arg.user, gid) < 0)
        {
	  /* non critical error */
          log_error ("couldn't add supplementary groups");
	}
    }

  /* Change root if requested */
  if (arg.root)
    {
      if (chroot (arg.root) < 0)
        {
          log_error ("couldn't change root");
          log_exit (1);
	}
      if (chdir ("/") < 0)
        {
          log_error ("couldn't change dir to new root");
          log_exit (1);
	}
      if (fclose (stdin) || fclose (stdout) || fclose (stderr))
        {
          log_error ("couldn't close stdin, stdout and/or stderr");
          log_exit (1);
	}
    }

  /* Change user if requested */
  if (arg.user)
    {
      if (setuid (uid) < 0)
        {
          log_error ("couldn't change user");
          log_exit (1);
	}
    }

  for (;;)
    {
      time_t last_tunnel_write;

      log_debug ("waiting for tunnel connection");

      if (arg.device != NULL)
	{
	  fd = open_device (arg.device);
	  log_debug ("open_device (\"%s\") = %d", arg.device, fd);
	  if (fd == -1)
	    {
	      log_error ("couldn't open %s: %s",
			 arg.device, strerror (errno));
	      log_exit (1);
	    }
	  /* Check that fd is not 0 (clash with --stdin-stdout) */
	  if (fd == 0)
	    {
	      log_notice ("changing fd from %d to 3", fd);
	      if (dup2 (fd, 3) != 3)
	        {
		  log_error ("couldn't dup2(%d, 3): %s", fd, strerror (errno));
		  log_exit (1);
		}
	    }
	} else if (arg.use_std) {
	  log_debug ("using stdin as fd");
	  fd = 0;
	  if (fcntl (fd, F_SETFL, O_NONBLOCK)==-1)
	    {
	      log_error ("couldn't set stdin to non-blocking mode: %s",
			 strerror (errno));
	      log_exit (1);
	    }
	  /* Usage of stdout (fd = 1) is checked later. */
	}

      if (tunnel_accept (tunnel) == -1)
	{
	  log_notice ("couldn't accept connection: %s", strerror (errno));
	  continue;
	}

      if (arg.forward_port != -1)
	{
	  struct sockaddr_in addr;

	  if (set_address (&addr, arg.forward_host, arg.forward_port) == -1)
	    {
	      log_error ("couldn't forward port to %s:%d: %s\n",
			 arg.forward_host, arg.forward_port, strerror (errno));
	      log_exit (1);
	    }

	  fd = do_connect (&addr);
	  log_debug ("do_connect (\"%s:%d\") = %d",
		 arg.forward_host, arg.forward_port, fd);
	  if (fd == -1)
	    {
	      log_error ("couldn't connect to %s:%d: %s\n",
			 arg.forward_host, arg.forward_port, strerror (errno));
	      log_exit (1);
	    }
	  /* Check that fd is not 0 (clash with --stdin-stdout) */
	  if (fd == 0)
	    {
	      log_notice ("changing fd from %d to 3", fd);
	      if (dup2 (fd, 3) != 3)
	        {
		  log_error ("couldn't dup2(%d, 3): %s", fd, strerror (errno));
		  log_exit (1);
		}
	    }
	}

      closed = FALSE;
      time (&last_tunnel_write);
      while (!closed)
	{
	  struct pollfd pollfd[2];
	  int timeout;
	  time_t t;
	  int n;
      
	  pollfd[0].fd = fd;
	  pollfd[0].events = POLLIN;
	  pollfd[1].fd = tunnel_pollin_fd (tunnel);
	  pollfd[1].events = POLLIN;

	  time (&t);
	  timeout = 1000 * (arg.keep_alive - (t - last_tunnel_write));
	  if (timeout < 0)
	    timeout = 0;

	  log_annoying ("poll () ...");
	  n = poll (pollfd, 2, timeout);
	  log_annoying ("... = %d", n);
	  if (n == -1)
	    {
	      log_error ("poll error: %s\n", strerror (errno));
	      log_exit (1);
	    }
	  else if (n == 0)
	    {
	      log_verbose ("poll() timed out");
	      tunnel_padding (tunnel, 1);
	      time (&last_tunnel_write);
	      continue;
	    }

	  log_annoying ("revents[0] = %x, revents[1] = %x, POLLIN = %x",
			pollfd[0].revents, pollfd[1].revents, POLLIN);
	  handle_input ("device or port", tunnel, fd, pollfd[0].revents,
			handle_device_input, &closed);
	  handle_input ("tunnel", tunnel, fd, pollfd[1].revents,
			handle_tunnel_input, &closed);

	  if (pollfd[0].revents & POLLIN)
	    time (&last_tunnel_write);
	}

      log_debug ("closing tunnel");
      if (fd != 0)
        {
          close (fd);
	}
      tunnel_close (tunnel);
      log_notice ("disconnected from FIXME:hostname:port");
    }

  log_debug ("destroying tunnel");
  tunnel_destroy (tunnel);
 
  log_exit (0);
}
